src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp

```diff
@@ -916,94 +867,52 @@ void ShenandoahHeapRegion::set_affiliation(ShenandoahRegionAffiliation new_affil
-
-    // Since the humongous region holds only one object, no lock is necessary for this register_object() invocation.
-    heap->card_scan()->register_object_wo_lock(bottom());
-    size_t index_limit = index() + ShenandoahHeapRegion::required_regions(obj->size() * HeapWordSize);
-
-    // For this region and each humongous continuation region spanned by this humongous object, change
-    // affiliation to OLD_GENERATION and adjust the generation-use tallies.  The remnant of memory
-    // in the last humongous region that is not spanned by obj is currently not used.
-    for (size_t i = index(); i < index_limit; i++) {
-      ShenandoahHeapRegion* r = heap->get_region(i);
-      log_debug(gc)("promoting region " SIZE_FORMAT ", from " SIZE_FORMAT " to " SIZE_FORMAT,
-        r->index(), (size_t) r->bottom(), (size_t) r->top());
-      if (r->top() < r->end()) {
-        ShenandoahHeap::fill_with_object(r->top(), (r->end() - r->top()) / HeapWordSize);
-        heap->card_scan()->register_object_wo_lock(r->top());
-        heap->card_scan()->mark_range_as_clean(top(), r->end() - r->top());
-      }
-      // We mark the entire humongous object's range as dirty after loop terminates, so no need to dirty the range here
-      r->set_affiliation(OLD_GENERATION);
-      log_debug(gc)("promoting humongous region " SIZE_FORMAT ", dirtying cards from " SIZE_FORMAT " to " SIZE_FORMAT,
-                    i, (size_t) r->bottom(), (size_t) r->top());
-      old_generation->increase_used(r->used());
-      young_generation->decrease_used(r->used());
-    }
-    if (promoting_all || obj->is_typeArray()) {
-      // Primitive arrays don't need to be scanned.  Likewise, if we are promoting_all, there's nothing
-      // left in young-gen, so there can exist no "interesting" pointers.  See above TODO question about requiring
-      // region promotion at safepoint.  If we're not at a safepoint, then we can't really "promote all" without
-      // directing new allocations to old-gen.  That's probably not what we want.  The whole "promote-all strategy"
-      // probably needs to be revisited at some future point.
-      heap->card_scan()->mark_range_as_clean(bottom(), obj->size());
-    } else {
-      heap->card_scan()->mark_range_as_dirty(bottom(), obj->size());
-    }
-    return index_limit - index();
+  oop obj = cast_to_oop(bottom());
+  assert(marking_context->is_marked(obj), "promoted humongous object should be alive");
+
+  size_t spanned_regions = ShenandoahHeapRegion::required_regions(obj->size() * HeapWordSize);
+  size_t index_limit = index() + spanned_regions;
+
+  log_debug(gc)("promoting humongous region " SIZE_FORMAT ", spanning " SIZE_FORMAT, index(), spanned_regions);
+
+  // Since this region may have served previously as OLD, it may hold obsolete object range info.
+  heap->card_scan()->reset_object_range(bottom(), bottom() + spanned_regions * ShenandoahHeapRegion::region_size_words());
```

Comment:
Do we write this memory again on line 911 or 915?

Reply:
The code at lines 911 and 915 clears or sets the card dirty bits.  The code here at line 890 changes the remembered set's knowledge of where objects begin.  These are two different activities.